### ISSUE TITLE
Unable to solve MethodCallExpr with enum paramter using ReflectionTypeSolver

### ISSUE THREAD

**Issue by HoldYourWaffle** [javaparser/javaparser#1942]
I'm trying to create a JUnit test that checks if all the translation keys I use in my project actually exist in the language definitions.
My (generalised) code is as follows:

```java
TypeSolver typeSolver = new ReflectionTypeSolver(false);
JavaSymbolSolver symbolSolver = new JavaSymbolSolver(typeSolver);
JavaParser.getStaticConfiguration().setSymbolResolver(symbolSolver);

for (CompilationUnit source : obtainSourcesOfProject()) {
    for (MethodCallExpr call : source.findAll(MethodCallExpr.class)) {
        // ... snip ...
        ResolvedMethodDeclaration dec = call.resolve();
        // ... snip ...
    }
}
```

This works for most method calls, but it chokes on the expression `something.equals(Enum.CONSTANT)`, throwing the following exception:
```
java.lang.RuntimeException: Unable to calculate the type of a parameter of a method call. Method call: something.equals(Enum.CONSTANT), Parameter: Enum.CONSTANT
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveArguments(JavaParserFacade.java:225)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solve(JavaParserFacade.java:240)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solve(JavaParserFacade.java:125)
	at com.github.javaparser.symbolsolver.JavaSymbolSolver.resolveDeclaration(JavaSymbolSolver.java:142)
	at com.github.javaparser.ast.expr.MethodCallExpr.resolve(MethodCallExpr.java:313)
	... snip ...
Caused by: java.lang.UnsupportedOperationException
	at com.github.javaparser.symbolsolver.reflectionmodel.ReflectionEnumConstantDeclaration.getType(ReflectionEnumConstantDeclaration.java:29)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.solveDotExpressionType(TypeExtractor.java:162)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:181)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:44)
	at com.github.javaparser.ast.expr.FieldAccessExpr.accept(FieldAccessExpr.java:90)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:447)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:310)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveArguments(JavaParserFacade.java:221)
	... 32 more
```

After a short investigation I found out that [`ReflectionEnumConstantDeclaration.getType`](https://github.com/javaparser/javaparser/blob/d5698c6b118d6741dd8d984c37091196b7a0a810/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/reflectionmodel/ReflectionEnumConstantDeclaration.java#L29) is unsupported.

Although this looks like a common scenario to me, I couldn't find a solution or anything resembling my issue anywhere on the internet. This mean that either nobody has done this before (which would surprise me), or I'm missing something very obvious.

This is my first project using JavaParser, so I'm almost certain I'm missing something.
I can't use a different `TypeSolver` because my project sits inside of a rather complicated gradle setup, so I can't just point it to a folder/jar with all my dependencies (unless there's a way to do that which I'm unaware of).
I'm not even interested in this particular expression, but I can't filter it out before resolving it. Is there maybe a way to 'resolve' the call expression without the parameter type information?

Other information that might be important:
- This code sits inside a JUnit test that's executed inside eclipse or the gradle command line
- The dependency situation of the project is rather complicated, so I can't use a different `TypeSolver` and point it to a folder with all my dependencies
- The enum that's failing to resolve is declared inside one of my dependencies

Any pointers on how to solve this would be greatly appreciated!

**Comment by un0btanium** [comment]
>The enum that's failing to resolve is declared inside one of my dependencies

That is probably why it fails ;)

> I can't use a different TypeSolver because my project sits inside of a rather complicated gradle setup, so I can't just point it to a folder/jar with all my dependencies (unless there's a way to do that which I'm unaware of).

You can use the `CombinedTypeSolver` to register your dependencies.
```
TypeSolver typeSolver = new CombinedTypeSolver(
	new ReflectionTypeSolver(),
	JarTypeSolver.getJarTypeSolver("jars/library1.jar"),
	JarTypeSolver.getJarTypeSolver("jars/library2.jar"),
	JarTypeSolver.getJarTypeSolver("jars/library3.jar"),
	new JavaParserTypeSolver(new File("src/main/java")),
);
```

You can also create a `CombinedTypeSolver` and add more JARs one-by-one to it:
```
TypeSolver typeSolver = new CombinedTypeSolver(
	new ReflectionTypeSolver(),
);
typeSolver.add(new JarTypeSolver("jars/library1.jar));
typeSolver.add(new JarTypeSolver("jars/library2.jar));
typeSolver.add(new JarTypeSolver("jars/library3.jar));
```

Since I am also using Gradle and ran into a similar problem, I solved it with JVM System Properties. You can send all required information from your gradle task to your java executable with them. In your task you write:
```
jvmArgs = ["-DprojectDependencies=${libraryJarPaths}"]
```
Of course you have to fill `libraryJarPaths` with the paths off your libraries manually/somehow.

In your program using JavaParser you extract that information via `System.getProperty("projectDependencies")`. Be aware that this returns a `String` and you first have to either cast it if you pass a list of paths or extract it with something like `String.split(", ")`.
Iterate over the list and wrap those paths in a `JarTypeSolver` and pass it to the `CombinedTypeSolver` with the `.add(TypeSolver)` method.

Hope this helps :)

**Comment by HoldYourWaffle** [comment]
Thank you for your response! It has helped me come further, but I encountered another error.

<br>

> *The enum that's failing to resolve is declared inside one of my dependencies*
That is probably why it fails ;)

That seems unlikely to me, as [`ReflectionEnumConstantDeclaration.getType`](https://github.com/javaparser/javaparser/blob/d5698c6b118d6741dd8d984c37091196b7a0a810/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/reflectionmodel/ReflectionEnumConstantDeclaration.java#L29) always throws an `UnsupportedOperationException`. Am I missing something here?

<br>

> Since I am also using Gradle and ran into a similar problem, I solved it with JVM System Properties. You can send all required information from your gradle task to your java executable with them. In your task you write:
`jvmArgs = ["-DprojectDependencies=${libraryJarPaths}"]`
...

This didn't solve my problem, but it did give me the idea of using system properties to find the dependencies. I was able to *solve* my problem using the following setup-code:
```java
List<String> paths = Arrays.asList(System.getProperty("java.class.path").split(System.getProperty("path.separator"))); //split the current classpath using the system path separator (';' on windows)
List<TypeSolver> solvers = paths.stream().filter(s->s.endsWith(".jar")).map(JarTypeSolver::getJarTypeSolver)).collect(Collectors.toList()); //this doesn't compile because of an unhandled IOException, used a method reference for clarity
solvers.add(new ReflectionTypeSolver(false)); //for JRE libraries

TypeSolver typeSolver = new CombinedTypeSolver(solvers.toArray(new TypeSolver[0]));
JavaSymbolSolver symbolSolver = new JavaSymbolSolver(typeSolver);
JavaParser.getStaticConfiguration().setSymbolResolver(symbolSolver);
```

This enables the solving of the original enum-parameter line, but now it chokes on a different line containing some generics-shenanigans. I'll create a new issue for this since it's unrelated to this one.

Thanks for your push in the right direction!

**Comment by HoldYourWaffle** [comment]
This isn't actually solved yet, just a workaround.

**Comment by ftomassetti** [comment]
Fixed by #1947
